sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIOReadFiles.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.csv;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.io.FileIO;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * Skeleton for error handling in CsvIO that transforms a {@link FileIO.ReadableFile} into the
+ * result of parsing.
+ */
+// TODO(https://github.com/apache/beam/issues/31736): Plan completion in future PR after
+// dependencies are completed.
+class CsvIOReadFiles<T> extends PTransform<PCollection<FileIO.ReadableFile>, CsvIOParseResult<T>> {
+  /** Stores required parameters for parsing. */
+  private final CsvIOParseConfiguration.Builder configBuilder;
```

Comment:
Usually the (mutable) Builder is an intermediate object and will be immediately converted to read-only configuration. Shall we store CsvIOParseConfiguration reference here instead? CC: @damondouglas

Reply:
Since configuration is package private, it doesn't matter and allows for the validation/check to occur at expand. This prevents a lot of redundant with methods and null checks.

Reply:
> This prevents a lot of redundant with methods and null checks.

I am not sure what this means. We can come back later when CsvIOReadFiles transform is building up.

Just some design comments for now -

Many Beam IOs follows this `.with...` methods pattern. It is verbose (as Java always be like).

also,
```
if (configuration.a != null
checkNotNull(configuration.b, "b must be provided if a is set")
```
which is basically the pattern of argument validation in various Beam IOs.

Do you mean CsvIOReadFiles will be structured differently than most Beam IO practices? I would generally expect new IO follows similar pattern in API so user has similar experience.